src/Compilers/Core/Portable/PEWriter/PeWriter.cs

```diff
@@ -242,6 +250,30 @@ internal static class PeWriter
             }
 
             return true;
+
+            Blob getChecksumBlob(PEBuilder builder)
+                => (Blob)typeof(PEBuilder).GetRuntimeFields()
+                    .Where(f => f.Name == "_lazyChecksum")
+                    .Single()
+                    .GetValue(builder);
+        }
+
+        private static MethodInfo _calculateChecksumMethod;
```

Comment:
I think we can correct the checksum without using reflection. Calculating the offset of the checksum in the PE file should be easy. Then we can copy the [checksum calculation implementation](http://source.dot.net/#System.Reflection.Metadata/System/Reflection/PortableExecutable/PEBuilder.cs,518) from SRM and:
1) build the full signed PE image as notmal
2) zero the checksum field
3) calculate the checksum 
4) write the checksum field

@agocke Wouldn't that work?

Reply:
Hashing 0-bytes is not a no-op. It's not good enough to zero the signature, you actually have to skip it.

Reply:
The signature is already calculated with the checksum zeroed (see the impl in SRM).

Reply:
Sorry, I wrote signature but I meant checksum. If I copied the code I would need to find the offset of the checksum field and exclude it from hashing. The APIs to do that are private, as far as I can tell. I could maybe get the raw byte array and then hard code the offset of the signature field, as well as copy over all the checksum code, but this doesn't seem better than using private reflection.

Reply:
The offset of the checksum field is a constant: http://source.dot.net/#System.Reflection.Metadata/System/Reflection/PortableExecutable/PEHeader.cs,252 for all PE files.

You can easily pass 2 blobs to the CalculateChecksum function: all content before that field and after that field.

Reply:
I don't think I can construct those blobs, because the constructor for a blob with an underlying buffer is internal.

Reply:
A blob is just an array segment.